File: backend/app/services/vault.py
"""
Credential vault: Fernet-encrypted secrets stored in PostgreSQL.

Encrypt/decrypt credentials at the application layer.
Master key from ROKA_VAULT_KEY env var -- never stored in DB.
"""

import json
import logging
import uuid
from typing import Any

from cryptography.fernet import Fernet, InvalidToken

from app.config import settings
from app.db import get_pool

logger = logging.getLogger(__name__)

_fernet: Fernet | None = None


def _get_fernet() -> Fernet:
    global _fernet
    if _fernet is None:
        key = settings.roka_vault_key.strip()
        if not key:
            raise RuntimeError(
                "ROKA_VAULT_KEY not set. Generate one with: "
                "python -c \"from cryptography.fernet import Fernet; print(Fernet.generate_key().decode())\""
            )
        _fernet = Fernet(key.encode())
    return _fernet


def encrypt(data: dict[str, Any]) -> bytes:
    """Encrypt a dict as Fernet-encrypted bytes."""
    payload = json.dumps(data).encode("utf-8")
    return _get_fernet().encrypt(payload)


def decrypt(encrypted: bytes) -> dict[str, Any]:
    """Decrypt Fernet-encrypted bytes back to a dict."""
    try:
        payload = _get_fernet().decrypt(encrypted)
        return json.loads(payload)
    except InvalidToken:
        logger.error("Failed to decrypt credential -- wrong vault key?")
        raise ValueError("Decryption failed. Check ROKA_VAULT_KEY.")


async def create_credential(
    owner_id: str,
    name: str,
    service: str,
    cred_type: str,
    config: dict[str, Any],
) -> dict[str, Any]:
    """Create a new encrypted credential. Returns the row (without decrypted config)."""
    pool = get_pool()
    encrypted = encrypt(config)
    row = await pool.fetchrow("""
        INSERT INTO credentials (owner_id, name, service, type, config_encrypted)
        VALUES ($1, $2, $3, $4::credential_type, $5)
        RETURNING id, name, service, type::text, is_active, created_at, updated_at
    """, uuid.UUID(owner_id), name, service, cred_type, encrypted)
    return dict(row) if row else {}


async def update_credential(
    credential_id: str,
    owner_id: str,
    name: str | None = None,
    service: str | None = None,
    config: dict[str, Any] | None = None,
    is_active: bool | None = None,
) -> dict[str, Any]:
    """Update credential fields. Config is re-encrypted if provided."""
    pool = get_pool()
    sets = []
    args: list[Any] = []
    idx = 3  # $1=id, $2=owner_id

    if name is not None:
        sets.append(f"name = ${idx}")
        args.append(name)
        idx += 1
    if service is not None:
        sets.append(f"service = ${idx}")
        args.append(service)
        idx += 1
    if config is not None:
        sets.append(f"config_encrypted = ${idx}")
        args.append(encrypt(config))
        idx += 1
    if is_active is not None:
        sets.append(f"is_active = ${idx}")
        args.append(is_active)
        idx += 1

    if not sets:
        return {}

    row = await pool.fetchrow(f"""
        UPDATE credentials SET {', '.join(sets)}
        WHERE id = $1 AND owner_id = $2
        RETURNING id, name, service, type::text, is_active, created_at, updated_at
    """, uuid.UUID(credential_id), uuid.UUID(owner_id), *args)
    return dict(row) if row else {}


async def get_credential_decrypted(credential_id: str) -> dict[str, Any]:
    """Fetch and decrypt a credential by ID. Backend-only -- never expose to frontend."""
    pool = get_pool()
    row = await pool.fetchrow(
        "SELECT id, owner_id, name, service, type::text, config_encrypted, is_active FROM credentials WHERE id = $1",
        uuid.UUID(credential_id),
    )
    if not row:
        raise ValueError(f"Credential {credential_id} not found")
    config = decrypt(row["config_encrypted"])
    return {
        "id": str(row["id"]),
        "owner_id": str(row["owner_id"]),
        "name": row["name"],
        "service": row["service"],
        "type": row["type"],
        "config": config,
        "is_active": row["is_active"],
    }


async def get_credentials_by_service(service: str, owner_id: str) -> list[dict[str, Any]]:
    """Fetch all active credentials for a service, decrypted."""
    pool = get_pool()
    rows = await pool.fetch("""
        SELECT id, name, service, type::text, config_encrypted, is_active
        FROM credentials
        WHERE service = $1 AND owner_id = $2 AND is_active = true
    """, service, uuid.UUID(owner_id))
    results = []
    for row in rows:
        config = decrypt(row["config_encrypted"])
        results.append({
            "id": str(row["id"]),
            "name": row["name"],
            "service": row["service"],
            "type": row["type"],
            "config": config,
        })
    return results


async def list_credentials(owner_id: str) -> list[dict[str, Any]]:
    """List credentials for a user (without decrypted config)."""
    pool = get_pool()
    rows = await pool.fetch("""
        SELECT id, name, service, type::text, is_active, created_at, updated_at
        FROM credentials
        WHERE owner_id = $1
        ORDER BY created_at DESC
    """, uuid.UUID(owner_id))
    return [dict(r) for r in rows]


async def delete_credential(credential_id: str, owner_id: str) -> bool:
    result = await get_pool().execute(
        "DELETE FROM credentials WHERE id = $1 AND owner_id = $2",
        uuid.UUID(credential_id), uuid.UUID(owner_id),
    )
    return result == "DELETE 1"

File: frontend/lib/vault.ts
/**
 * Fernet-compatible credential vault.
 * Shared by both the Next.js API routes (OAuth exchange) and the worker process.
 *
 * Uses the same ROKA_VAULT_KEY (32-byte base64 Fernet key) as the Python backend
 * so existing encrypted credentials remain readable.
 */

import { Secret, Token } from "fernet";
import { sql } from "drizzle-orm";
import type { PostgresJsDatabase } from "drizzle-orm/postgres-js";
import type * as schema from "@/lib/db/schema";

const OAUTH_REFRESH_BUFFER_SEC = 300; // refresh 5 min before expiry

// ---------------------------------------------------------------------------
// Core encrypt / decrypt
// ---------------------------------------------------------------------------

function getSecret(): Secret {
  const key = process.env.ROKA_VAULT_KEY;
  if (!key) throw new Error("ROKA_VAULT_KEY not set");
  return new Secret(key);
}

export function encrypt(data: Record<string, unknown>): string {
  const secret = getSecret();
  const token = new Token({ secret });
  return token.encode(JSON.stringify(data));
}

export function decrypt(encrypted: string): Record<string, unknown> {
  const secret = getSecret();
  const token = new Token({ secret, token: encrypted, ttl: 0 });
  return JSON.parse(token.decode()) as Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Credential CRUD
// ---------------------------------------------------------------------------

export interface CredentialRow {
  id: string;
  ownerId: string;
  name: string;
  service: string;
  type: string;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export async function createCredential(
  db: PostgresJsDatabase<typeof schema>,
  opts: {
    ownerId: string;
    name: string;
    service: string;
    type: string;
    config: Record<string, unknown>;
  },
): Promise<CredentialRow> {
  const enc = encrypt(opts.config);
  const rows = await db.execute(sql`
    INSERT INTO credentials (owner_id, name, service, type, config_encrypted)
    VALUES (${opts.ownerId}::uuid, ${opts.name}, ${opts.service}, ${opts.type}::credential_type, ${enc})
    RETURNING id, owner_id AS "ownerId", name, service, type, is_active AS "isActive",
              created_at AS "createdAt", updated_at AS "updatedAt"
  `);
  return rows[0] as unknown as CredentialRow;
}

export async function getCredentialDecrypted(
  db: PostgresJsDatabase<typeof schema>,
  credentialId: string,
): Promise<{ id: string; service: string; type: string; config: Record<string, unknown> }> {
  const rows = await db.execute(sql`
    SELECT id, service, type, config_encrypted AS "configEncrypted"
    FROM credentials WHERE id = ${credentialId}::uuid
  `);
  if (!rows[0]) throw new Error(`Credential ${credentialId} not found`);
  const row = rows[0] as unknown as { id: string; service: string; type: string; configEncrypted: string };
  return { id: row.id, service: row.service, type: row.type, config: decrypt(row.configEncrypted) };
}

export async function getCredentialsByService(
  db: PostgresJsDatabase<typeof schema>,
  service: string,
  ownerId: string,
): Promise<Array<{ id: string; service: string; type: string }>> {
  const rows = await db.execute(sql`
    SELECT id, service, type FROM credentials
    WHERE service = ${service} AND owner_id = ${ownerId}::uuid AND is_active = true
    ORDER BY updated_at DESC
  `);
  return rows as unknown as Array<{ id: string; service: string; type: string }>;
}

// ---------------------------------------------------------------------------
// OAuth token refresh
// ---------------------------------------------------------------------------

function isOAuthExpired(config: Record<string, unknown>): boolean {
  const expiresAt = config.expires_at as number | undefined;
  if (!expiresAt) return false;
  return Date.now() / 1000 > expiresAt - OAUTH_REFRESH_BUFFER_SEC;
}

async function refreshGoogleToken(config: Record<string, unknown>): Promise<Record<string, unknown>> {
  const res = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: config.refresh_token as string,
      client_id: config.client_id as string || process.env.GOOGLE_OAUTH_CLIENT_ID || "",
      client_secret: config.client_secret as string || process.env.GOOGLE_OAUTH_CLIENT_SECRET || "",
    }),
  });
  if (!res.ok) throw new Error(`Google token refresh failed: ${res.status}`);
  const data = await res.json();
  return {
    ...config,
    access_token: data.access_token,
    expires_at: Math.floor(Date.now() / 1000) + (data.expires_in || 3600),
  };
}

async function refreshSlackToken(config: Record<string, unknown>): Promise<Record<string, unknown>> {
  const res = await fetch("https://slack.com/api/oauth.v2.access", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: config.refresh_token as string,
      client_id: config.client_id as string || process.env.SLACK_OAUTH_CLIENT_ID || "",
      client_secret: config.client_secret as string || process.env.SLACK_OAUTH_CLIENT_SECRET || "",
    }),
  });
  if (!res.ok) throw new Error(`Slack token refresh failed: ${res.status}`);
  const data = await res.json();
  if (!data.ok) throw new Error(`Slack error: ${data.error}`);
  return {
    ...config,
    access_token: data.access_token,
    refresh_token: data.refresh_token || config.refresh_token,
    expires_at: data.expires_in ? Math.floor(Date.now() / 1000) + data.expires_in : config.expires_at,
  };
}

/**
 * Fetches and returns a valid access_token for a credential.
 * Refreshes automatically if expired.
 */
export async function ensureValidToken(
  db: PostgresJsDatabase<typeof schema>,
  credentialId: string,
  ownerId: string,
): Promise<string> {
  const cred = await getCredentialDecrypted(db, credentialId);
  const config = cred.config;

  if (cred.type !== "oauth2") {
    return (config.api_key as string) || (config.access_token as string) || "";
  }

  if (!isOAuthExpired(config)) {
    return config.access_token as string;
  }

  let refreshed: Record<string, unknown>;
  if (cred.service === "google") {
    refreshed = await refreshGoogleToken(config);
  } else if (cred.service === "slack") {
    refreshed = await refreshSlackToken(config);
  } else {
    return config.access_token as string;
  }

  const enc = encrypt(refreshed);
  await db.execute(sql`
    UPDATE credentials SET config_encrypted = ${enc}, updated_at = now()
    WHERE id = ${credentialId}::uuid
  `);

  return refreshed.access_token as string;
}
